cli/parse.go

```diff
@@ -14,265 +15,182 @@ import (
 	"github.com/ipfs/go-ipfs-cmdkit/files"
 	"github.com/ipfs/go-ipfs-cmds"
 
-	u "github.com/ipfs/go-ipfs-util"
 	logging "github.com/ipfs/go-log"
 )
 
 var log = logging.Logger("cmds/cli")
 
 // Parse parses the input commandline string (cmd, flags, and args).
 // returns the corresponding command Request object.
-func Parse(input []string, stdin *os.File, root *cmds.Command) (cmds.Request, *cmds.Command, []string, error) {
-	path, opts, stringVals, cmd, err := parseOpts(input, root)
+func Parse(ctx context.Context, input []string, stdin *os.File, root *cmds.Command) (*cmds.Request, error) {
+	req, err := parse(input, root)
 	if err != nil {
-		return nil, nil, path, err
+		return req, err
 	}
 
-	optDefs, err := root.GetOptions(path)
-	if err != nil {
-		return nil, cmd, path, err
-	}
-
-	req, err := cmds.NewRequest(path, opts, nil, nil, cmd, optDefs)
-	if err != nil {
-		return nil, cmd, path, err
-	}
+	req.Context = ctx
 
 	// This is an ugly hack to maintain our current CLI interface while fixing
 	// other stdin usage bugs. Let this serve as a warning, be careful about the
 	// choices you make, they will haunt you forever.
-	if len(path) == 2 && path[0] == "bootstrap" {
-		if (path[1] == "add" && opts["default"] == true) ||
-			(path[1] == "rm" && opts["all"] == true) {
+	if len(req.Path) == 2 && req.Path[0] == "bootstrap" {
+		if (req.Path[1] == "add" && req.Options["default"] == true) ||
+			(req.Path[1] == "rm" && req.Options["all"] == true) {
 			stdin = nil
 		}
 	}
 
-	stringArgs, fileArgs, err := ParseArgs(req, stringVals, stdin, cmd.Arguments, root)
-	if err != nil {
-		return req, cmd, path, err
+	if err = parseArgs(req, root, stdin); err != nil {
```

Comment:
style thing, but if i'm using this pattern, I always make sure to use `if err := ` so we don't accidentally reuse the outer scoped variable

Reply:
What's wrong with reusing that variable? All we do is set it right before we return.